File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:    
    def maxProfit(self, prices: List[int]) -> int:
        buyIndex = 0 # first day
        sellIndex = 1 # second day
        maxProfit = 0
        while sellIndex < len(prices):
            if prices[buyIndex] < prices[sellIndex]:
                profit = prices[sellIndex] - prices[buyIndex]
                maxProfit = max(profit, maxProfit)
            else:
                buyIndex = sellIndex
            sellIndex += 1
        return maxProfit

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.rs
use std::cmp::{min, max};

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max_profit = 0;
        let mut min_price = prices[0];
        
        for price in prices {
            min_price = min(price, min_price);
            let current_profit = price - min_price;
            max_profit = max(current_profit, max_profit);
        }
        
        return max_profit; // could omit `return`
    }
}
